README.md
ReflectorDispatcher implements [Dispatcher] by introspecting a class instance
using [package:reflectable]. package:reflectable allows you to invoke an instance's methods
by their string names.

Construct a [ReflectorDispatcher] with a class instance and its pre-built introspection data created when you build according to the directions in Reflectable.
ReflectorDispatcher.dispatch("someMethod") will return a Future of whatever
value it returns.

This is for JSON-RPC, so there are some restrictions. A method can only be invoked with 
positional or named parameters, not both. Optional parameters should work as expected. 

ReflectorDispatcher is mainly a wrapper around package:reflectable, as a replacement
for dart:mirrors, which is incompatible with Flutter.
For any method dispatched, Dispatcher returns either the return value of
the method or instances of one of three "Exception" classes. Most errors
will be corralled into these objects, so that runtime exceptions don't get
thrown, instead are returned in an orderly manner.

## Usage

A simple usage example:

```dart
import 'package:reflector_dispatcher/reflector_dispatcher.dart';
import 'package:reflectable/reflectable.dart';

/// this import does not exist until you run the builder
/// (presuming this is in an 'example' folder)
/// > pub run build_runner build example
/// or > flutter pub run build_runner build example
import 'reflector_dispatcher_example.reflectable.dart';

// Annotate with this class to enable reflection.
class Reflector extends Reflectable {
  const Reflector()
      : super(
            invokingCapability, // Request the capability to invoke methods.
            declarationsCapability); // needed for introspecting methods.
}

const reflectable = Reflector();

/// Look in test/reflector_dispatcher_test.dart for additional examples.

/// Make a class with the API we want. Something silly but instructive.
///
/// Hold count in Foo.
@reflectable
class Foo {
  num _count = 0;

  /// initialize with a count
  Foo(this._count);

  /// increment with an optional positional parameter
  void increment([num aNumber = 1]) => _count += aNumber;

  /// decrement with a required named parameter
  void decrement({required num aNumber}) => _count -= aNumber;

  /// get the current value of count
  num getCount() => _count;

  /// we decrement by 2 a lot.
  num goTwoLess() {
    _count -= 2;
    return _count;
  }
}

void main() async {
  // do this before using the ReflectorDispatcher
  initializeReflectable();

  /// make a dispatcher with an initialized instance of Foo class and its
  /// introspected data
  var dispatcher = ReflectorDispatcher(Foo(29), reflectable);

  /// dispatch a method with a parameter
  await dispatcher.dispatch('increment', 4);

  /// get the new value
  var c = await (dispatcher.dispatch('getCount'));
  printCount(c); // 33

  /// dispatch a method with a named parameter
  await dispatcher.dispatch('decrement', {'aNumber': 18});
  var d = await (dispatcher.dispatch('getCount'));
  printCount(d); // 15

  /// dispatch a method without a parameter
  await dispatcher.dispatch('increment');
  var e = await (dispatcher.dispatch('getCount'));
  printCount(e); // 16

  var f = await dispatcher.dispatch('goTwoLess');
  printCount(f); // 14
}

void printCount(num aCount) {
  print('the currentCount is $aCount');
}

```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/jwashin/reflector_dispatcher


CHANGELOG.md
## 1.0.0

- Initial version, created by Stagehand

## 1.0.1

- updated repository url

test/reflector_dispatcher_test.dart
@TestOn('vm')

import 'package:reflectable/reflectable.dart';
import 'package:reflector_dispatcher/reflector_dispatcher.dart';
import 'package:rpc_dispatcher/rpc_dispatcher.dart';
import 'package:rpc_exceptions/rpc_exceptions.dart';
import 'package:test/test.dart';

// the following file does not exist until generated by build_runner.
//
// > pub run build_runner build test
// or > flutter pub run build_runner build test
import 'reflector_dispatcher_test.reflectable.dart';

// Annotate with this class to enable reflection.
class Reflector extends Reflectable {
  const Reflector()
      : super(
            invokingCapability, // Request the capability to invoke methods.
            declarationsCapability); // needed for introspecting methods.
}

const introspected = Reflector();

@introspected
class Foo {
  String greetName;
  Foo([this.greetName = 'Stranger']);
  String hi() => 'Hi!';
  String hello() => 'Hello, $greetName!';
  String greet([String? name]) {
    var tmp = name ?? greetName;
    if (tmp == greetName) {
      return 'Hello, $tmp!';
    }
    return 'Hi, $tmp!';
  }

  num add(num a, num b) => a + b;
  num _privateAdd(num a, num b) => a + b;
  num subtract(num a, num b) => a - b;
  num subtractNamed({required num minuend, required num subtrahend}) =>
      minuend - subtrahend;
  dynamic throwError(num a, num b) {
    throw Zerr('you expected this!');
  }

  dynamic typeerror(dynamic a) {
    try {
      return a + 9;
    } on TypeError {
      throw RuntimeException('Cannot add string and number', -22, [a, 9]);
    }
  }

  num divzerotest(num a) {
    return a / 0;
  }

  num usePrivateAdd(num a, num b) {
    return _privateAdd(a, b);
  }
}

class Zerr implements Exception {
  String message;
  Zerr(this.message);
}

void main() {
  initializeReflectable();
  test('symbolize', () {
    expect(symbolizeKeys({'a': 1}), equals({Symbol('a'): 1}));
  });

  test('simple object', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('hi').then((dynamic value) => expect(value, equals('Hi!')));
  });

  test('simple object with param', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('greet', ['Mary']).then(
        (dynamic value) => expect(value, equals('Hi, Mary!')));
  });

  test('simple object initialized with param unused', () {
    var z = ReflectorDispatcher(Foo('Bar'), introspected);
    z.dispatch('greet', ['Mary']).then(
        (dynamic value) => expect(value, equals('Hi, Mary!')));
  });

  test('simple object initialized with param used', () {
    var z = ReflectorDispatcher(Foo('Bob'), introspected);
    z
        .dispatch('hello')
        .then((dynamic value) => expect(value, equals('Hello, Bob!')));
  });

  test('simple object without optional param', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z
        .dispatch('greet')
        .then((dynamic value) => expect(value, equals('Hello, Stranger!')));
  });

  test('simple addition', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('add', [3, 4]).then((dynamic value) => expect(value, equals(7)));
  });

  test('simple subtraction a b', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('subtract', [42, 23]).then(
        (dynamic value) => expect(value, equals(19)));
  });
  test('simple subtraction b a', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch(
        'subtract', [23, 42]).then((var value) => expect(value, equals(-19)));
  });

  test('named subtraction in order', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('subtractNamed', [], {'minuend': 23, 'subtrahend': 42}).then(
        (var value) => expect(value, equals(-19)));
  });

  test('named subtraction out of order', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('subtractNamed', [], {'subtrahend': 42, 'minuend': 23}).then(
        (var value) => expect(value, equals(-19)));
  });

  test('mixed nums', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('add', [3, 4.3]).then((var value) => expect(value, equals(7.3)));
  });

  test('method not found', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('zadd', [3, 4.3]).then(
        (dynamic value) => expect(value, TypeMatcher<MethodNotFoundException>()));

    ///#new isInstanceOf<MethodNotFound>()));
  });

  test('private method call', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('_privateAdd', [3, 4.3]).then(
        (dynamic value) => expect(value, TypeMatcher<MethodNotFoundException>()));
  });

  test('invalid parameters too many', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('add', [3, 5, 8]).then(

        // (dynamic value) => expect(value, TypeMatcher<InvalidParameters>()));
        (dynamic value) => expect(value, TypeMatcher<InvalidParametersException>()));
  });

  test('invalid parameters bad value', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('add', [3, 'hello']).then(
        (dynamic value) => expect(value, TypeMatcher<InvalidParametersException>()));
  });

  test('invalid parameters too few', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('add', [3]).then(
        (dynamic value) => expect(value, TypeMatcher<InvalidParametersException>()));
  });

  test('internal error', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('throwError', [3, 0]).then(
        (dynamic value) => expect(value, TypeMatcher<RuntimeException>()));
  });

  test('private method invocation', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('_private_add', [3, 4.3]).then(
        (dynamic value) => expect(value, TypeMatcher<MethodNotFoundException>()));
  });

  test('attempt property invocation', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z
        .dispatch('greet_name')
        .then((dynamic value) => expect(value, TypeMatcher<MethodNotFoundException>()));
  });

  test('catch TypeError in application code', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('typeerror', ['a']).then(
        (dynamic value) => expect(value, TypeMatcher<RuntimeException>()));
  });

  test('divide by zero', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch('divzerotest', [3]).then(
        (dynamic value) => expect(value, double.infinity));
  });

  test('zero over zero', () {
    var z = ReflectorDispatcher(Foo(), introspected);
    z.dispatch(
        'divzerotest', [0]).then((var value) => expect(value.isNaN, true));
  });
}


test/reflector_dispatcher_test.reflectable.dart
// This file has been generated by the reflectable package.
// https://github.com/dart-lang/reflectable.
// @dart = 2.12

import 'dart:core';
import 'reflector_dispatcher_test.dart' as prefix0;

// ignore_for_file: prefer_adjacent_string_concatenation
// ignore_for_file: prefer_collection_literals
// ignore_for_file: unnecessary_const
// ignore_for_file: implementation_imports

// ignore:unused_import
import 'package:reflectable/mirrors.dart' as m;
// ignore:unused_import
import 'package:reflectable/src/reflectable_builder_based.dart' as r;
// ignore:unused_import
import 'package:reflectable/reflectable.dart' as r show Reflectable;

final _data = <r.Reflectable, r.ReflectorData>{
  const prefix0.Reflector(): r.ReflectorData(
      <m.TypeMirror>[
        r.NonGenericClassMirrorImpl(
            r'Foo',
            r'.Foo',
            7,
            0,
            const prefix0.Reflector(),
            const <int>[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 13],
            const <int>[
              14,
              15,
              16,
              17,
              18,
              1,
              2,
              3,
              4,
              5,
              6,
              7,
              8,
              9,
              10,
              11,
              12
            ],
            const <int>[],
            -1,
            {},
            {},
            {
              r'': (bool b) => ([greetName = 'Stranger']) =>
                  b ? prefix0.Foo(greetName) : null
            },
            -1,
            -1,
            const <int>[-1],
            null,
            null)
      ],
      <m.DeclarationMirror>[
        r.VariableMirrorImpl(r'greetName', 32773, 0, const prefix0.Reflector(),
            -1, -1, -1, null, null),
        r.MethodMirrorImpl(r'hi', 131074, 0, -1, -1, -1, null, const <int>[],
            const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'hello', 131074, 0, -1, -1, -1, null, const <int>[],
            const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'greet', 131074, 0, -1, -1, -1, null,
            const <int>[0], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'add', 131074, 0, -1, -1, -1, null,
            const <int>[1, 2], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'subtract', 131074, 0, -1, -1, -1, null,
            const <int>[3, 4], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'subtractNamed', 131074, 0, -1, -1, -1, null,
            const <int>[5, 6], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'throwError', 65538, 0, -1, -1, -1, null,
            const <int>[7, 8], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'typeerror', 65538, 0, -1, -1, -1, null,
            const <int>[9], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'divzerotest', 131074, 0, -1, -1, -1, null,
            const <int>[10], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'usePrivateAdd', 131074, 0, -1, -1, -1, null,
            const <int>[11, 12], const prefix0.Reflector(), null),
        r.ImplicitGetterMirrorImpl(const prefix0.Reflector(), 0, 11),
        r.ImplicitSetterMirrorImpl(const prefix0.Reflector(), 0, 12),
        r.MethodMirrorImpl(r'', 0, 0, -1, -1, -1, null, const <int>[13],
            const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'==', 131074, -1, -1, -1, -1, null, const <int>[15],
            const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'toString', 131074, -1, -1, -1, -1, null,
            const <int>[], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'noSuchMethod', 65538, -1, -1, -1, -1, null,
            const <int>[16], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'hashCode', 131075, -1, -1, -1, -1, null,
            const <int>[], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'runtimeType', 131075, -1, -1, -1, -1, null,
            const <int>[], const prefix0.Reflector(), null)
      ],
      <m.ParameterMirror>[
        r.ParameterMirrorImpl(r'name', 36870, 3, const prefix0.Reflector(), -1,
            -1, -1, null, null, null, null),
        r.ParameterMirrorImpl(r'a', 32774, 4, const prefix0.Reflector(), -1, -1,
            -1, null, null, null, null),
        r.ParameterMirrorImpl(r'b', 32774, 4, const prefix0.Reflector(), -1, -1,
            -1, null, null, null, null),
        r.ParameterMirrorImpl(r'a', 32774, 5, const prefix0.Reflector(), -1, -1,
            -1, null, null, null, null),
        r.ParameterMirrorImpl(r'b', 32774, 5, const prefix0.Reflector(), -1, -1,
            -1, null, null, null, null),
        r.ParameterMirrorImpl(r'minuend', 40966, 6, const prefix0.Reflector(),
            -1, -1, -1, null, null, null, #minuend),
        r.ParameterMirrorImpl(
            r'subtrahend',
            40966,
            6,
            const prefix0.Reflector(),
            -1,
            -1,
            -1,
            null,
            null,
            null,
            #subtrahend),
        r.ParameterMirrorImpl(r'a', 32774, 7, const prefix0.Reflector(), -1, -1,
            -1, null, null, null, null),
        r.ParameterMirrorImpl(r'b', 32774, 7, const prefix0.Reflector(), -1, -1,
            -1, null, null, null, null),
        r.ParameterMirrorImpl(r'a', 16390, 8, const prefix0.Reflector(), -1, -1,
            -1, null, null, null, null),
        r.ParameterMirrorImpl(r'a', 32774, 9, const prefix0.Reflector(), -1, -1,
            -1, null, null, null, null),
        r.ParameterMirrorImpl(r'a', 32774, 10, const prefix0.Reflector(), -1,
            -1, -1, null, null, null, null),
        r.ParameterMirrorImpl(r'b', 32774, 10, const prefix0.Reflector(), -1,
            -1, -1, null, null, null, null),
        r.ParameterMirrorImpl(
            r'greetName',
            38918,
            13,
            const prefix0.Reflector(),
            -1,
            -1,
            -1,
            null,
            null,
            'Stranger',
            null),
        r.ParameterMirrorImpl(r'_greetName', 32870, 12,
            const prefix0.Reflector(), -1, -1, -1, null, null, null, null),
        r.ParameterMirrorImpl(r'other', 32774, 14, const prefix0.Reflector(),
            -1, -1, -1, null, null, null, null),
        r.ParameterMirrorImpl(r'invocation', 32774, 16,
            const prefix0.Reflector(), -1, -1, -1, null, null, null, null)
      ],
      <Type>[prefix0.Foo],
      1,
      {
        r'==': (dynamic instance) => (x) => instance == x,
        r'toString': (dynamic instance) => instance.toString,
        r'noSuchMethod': (dynamic instance) => instance.noSuchMethod,
        r'hashCode': (dynamic instance) => instance.hashCode,
        r'runtimeType': (dynamic instance) => instance.runtimeType,
        r'hi': (dynamic instance) => instance.hi,
        r'hello': (dynamic instance) => instance.hello,
        r'greet': (dynamic instance) => instance.greet,
        r'add': (dynamic instance) => instance.add,
        r'subtract': (dynamic instance) => instance.subtract,
        r'subtractNamed': (dynamic instance) => instance.subtractNamed,
        r'throwError': (dynamic instance) => instance.throwError,
        r'typeerror': (dynamic instance) => instance.typeerror,
        r'divzerotest': (dynamic instance) => instance.divzerotest,
        r'usePrivateAdd': (dynamic instance) => instance.usePrivateAdd,
        r'greetName': (dynamic instance) => instance.greetName
      },
      {r'greetName=': (dynamic instance, value) => instance.greetName = value},
      null,
      [])
};

final _memberSymbolMap = null;

void initializeReflectable() {
  r.data = _data;
  r.memberSymbolMap = _memberSymbolMap;
}


lib/src/reflector_dispatcher_base.dart
/// [Dispatcher] is used as a kind of sneaky way of telling an instantiated
/// object to do something. Instead of invoking a method directly in code, you
/// give the dispatcher the instance and directions (method name and parameters)
/// about invoking the method, and the Dispatcher gets the instance to perform
/// the method and gives you the returned value.

import 'package:rpc_dispatcher/rpc_dispatcher.dart';
import 'package:rpc_exceptions/rpc_exceptions.dart';

/// Dispatcher introspects a class instance so you can invoke its methods by
/// their string names,
///
/// Dispatcher.dispatch("someMethod") will return a Future of whatever value it
/// returns.
///
/// ReflectorDispatcher is mainly a wrapper around reflectable. When you
/// process a Reflectable class, a mirror of the class's schematics is created.
/// A ReflectorDispatcher is inititalized with that and an actual instance.
/// Note that we don't actually pubspec depend on Reflectable here. This
/// Dispatcher works with the results of using Reflectable to introspect a
/// class.
class ReflectorDispatcher implements Dispatcher {
  /// the initialized class instance we will be invoking methods on.
  dynamic instance;

  /// an introspected reflection of the instance's class's guts,
  /// predigested by reflectable
  dynamic mirror;

  /// constructor
  ReflectorDispatcher(this.instance, this.mirror);

  ///  Invoke named method with parameters on the instance and
  /// return a Future of the result, if possible.
  ///
  ///  Catch, repackage and return (not throw or rethrow) *All Errors*.
  ///  positionalParams should be a List or null.
  ///  namedParams should be a Map of String:value or null.
  @override
  Future dispatch(String methodName,
      [dynamic positionalParams, Map<String, dynamic>? namedParams]) async {
    namedParams = namedParams ?? <String, dynamic>{};
    var posParams = positionalParams ?? [];
    if (posParams is Map<String, dynamic>) {
      namedParams = positionalParams;
      posParams = [];
    }

    if (posParams is! List) {
      posParams = [posParams];
    }
    namedParams = namedParams ?? <String, dynamic>{};
    var symbolMap = <Symbol, dynamic>{};
    if (namedParams.isNotEmpty) {
      symbolMap = symbolizeKeys(namedParams);
    }

    var instanceMirror = mirror.reflect(instance);

    // error check the method and parameter counts
    var methodSchemas = instanceMirror.type.declarations;
    if (methodSchemas.containsKey(methodName)) {
      var params = methodSchemas[methodName].parameters;
      var namedCount = 0;
      var optionals = 0;
      for (var p in params) {
        if (p.isNamed) {
          namedCount += 1;
        }
        if (p.isOptional) {
          optionals += 1;
        }
      }
      var actualCount = posParams.length;
      var requiredCount = params.length - namedCount - optionals;
      if (actualCount < requiredCount) {
        return InvalidParametersException('too few params');
      }
      if (actualCount > requiredCount + optionals) {
        return InvalidParametersException('too many params');
      }
    } else {
      return MethodNotFoundException('Not found: $methodName');
    }

    var resp;
    try {
      resp = await instanceMirror.invoke(methodName, posParams, symbolMap);
    } on TypeError catch (e) {
      return InvalidParametersException('$e');
    } catch (e) {
      // passthrough for custom user-correctable errors
      if (e is RuntimeException) {
        return e;
      }
      return RuntimeException('$e');
    }
    return resp;
  }
}


example/reflector_dispatcher_example.dart
import 'package:reflector_dispatcher/reflector_dispatcher.dart';
import 'package:reflectable/reflectable.dart';

/// this import does not exist until you run the builder
/// (presuming this is in an 'example' folder)
/// > pub run build_runner build example
/// or > flutter pub run build_runner build example
import 'reflector_dispatcher_example.reflectable.dart';

// Annotate with this class to enable reflection.
class Reflector extends Reflectable {
  const Reflector()
      : super(
            invokingCapability, // Request the capability to invoke methods.
            declarationsCapability); // needed for introspecting methods.
}

const reflectable = Reflector();

/// Look in test/reflector_dispatcher_test.dart for additional examples.

/// Make a class with the API we want. Something silly but instructive.
///
/// Hold count in Foo.
@reflectable
class Foo {
  num _count = 0;

  /// initialize with a count
  Foo(this._count);

  /// increment with an optional positional parameter
  void increment([num aNumber = 1]) => _count += aNumber;

  /// decrement with a required named parameter
  void decrement({required num aNumber}) => _count -= aNumber;

  /// get the current value of count
  num getCount() => _count;

  /// we decrement by 2 a lot.
  num goTwoLess() {
    _count -= 2;
    return _count;
  }
}

void main() async {
  // do this before using the ReflectorDispatcher
  initializeReflectable();

  /// make a dispatcher with an initialized instance of Foo class and its
  /// introspected data
  var dispatcher = ReflectorDispatcher(Foo(29), reflectable);

  /// dispatch a method with a parameter
  await dispatcher.dispatch('increment', 4);

  /// get the new value
  var c = await (dispatcher.dispatch('getCount'));
  printCount(c); // 33

  /// dispatch a method with a named parameter
  await dispatcher.dispatch('decrement', {'aNumber': 18});
  var d = await (dispatcher.dispatch('getCount'));
  printCount(d); // 15

  /// dispatch a method without a parameter
  await dispatcher.dispatch('increment');
  var e = await (dispatcher.dispatch('getCount'));
  printCount(e); // 16

  var f = await dispatcher.dispatch('goTwoLess');
  printCount(f); // 14
}

void printCount(num aCount) {
  print('the currentCount is $aCount');
}


example/reflector_dispatcher_example.reflectable.dart
// This file has been generated by the reflectable package.
// https://github.com/dart-lang/reflectable.
// @dart = 2.12

import 'dart:core';
import 'reflector_dispatcher_example.dart' as prefix0;

// ignore_for_file: prefer_adjacent_string_concatenation
// ignore_for_file: prefer_collection_literals
// ignore_for_file: unnecessary_const
// ignore_for_file: implementation_imports

// ignore:unused_import
import 'package:reflectable/mirrors.dart' as m;
// ignore:unused_import
import 'package:reflectable/src/reflectable_builder_based.dart' as r;
// ignore:unused_import
import 'package:reflectable/reflectable.dart' as r show Reflectable;

final _data = <r.Reflectable, r.ReflectorData>{
  const prefix0.Reflector(): r.ReflectorData(
      <m.TypeMirror>[
        r.NonGenericClassMirrorImpl(
            r'Foo',
            r'.Foo',
            7,
            0,
            const prefix0.Reflector(),
            const <int>[0, 1, 2, 3, 4],
            const <int>[5, 6, 7, 8, 9, 0, 1, 2, 3],
            const <int>[],
            -1,
            {},
            {},
            {r'': (bool b) => (_count) => b ? prefix0.Foo(_count) : null},
            -1,
            -1,
            const <int>[-1],
            null,
            null)
      ],
      <m.DeclarationMirror>[
        r.MethodMirrorImpl(r'increment', 262146, 0, -1, -1, -1, null,
            const <int>[0], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'decrement', 262146, 0, -1, -1, -1, null,
            const <int>[1], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'getCount', 131074, 0, -1, -1, -1, null,
            const <int>[], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'goTwoLess', 131074, 0, -1, -1, -1, null,
            const <int>[], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'', 0, 0, -1, -1, -1, null, const <int>[2],
            const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'==', 131074, -1, -1, -1, -1, null, const <int>[3],
            const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'toString', 131074, -1, -1, -1, -1, null,
            const <int>[], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'noSuchMethod', 65538, -1, -1, -1, -1, null,
            const <int>[4], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'hashCode', 131075, -1, -1, -1, -1, null,
            const <int>[], const prefix0.Reflector(), null),
        r.MethodMirrorImpl(r'runtimeType', 131075, -1, -1, -1, -1, null,
            const <int>[], const prefix0.Reflector(), null)
      ],
      <m.ParameterMirror>[
        r.ParameterMirrorImpl(r'aNumber', 38918, 0, const prefix0.Reflector(),
            -1, -1, -1, null, null, 1, null),
        r.ParameterMirrorImpl(r'aNumber', 40966, 1, const prefix0.Reflector(),
            -1, -1, -1, null, null, null, #aNumber),
        r.ParameterMirrorImpl(r'_count', 32806, 4, const prefix0.Reflector(),
            -1, -1, -1, null, null, null, null),
        r.ParameterMirrorImpl(r'other', 32774, 5, const prefix0.Reflector(), -1,
            -1, -1, null, null, null, null),
        r.ParameterMirrorImpl(r'invocation', 32774, 7,
            const prefix0.Reflector(), -1, -1, -1, null, null, null, null)
      ],
      <Type>[prefix0.Foo],
      1,
      {
        r'==': (dynamic instance) => (x) => instance == x,
        r'toString': (dynamic instance) => instance.toString,
        r'noSuchMethod': (dynamic instance) => instance.noSuchMethod,
        r'hashCode': (dynamic instance) => instance.hashCode,
        r'runtimeType': (dynamic instance) => instance.runtimeType,
        r'increment': (dynamic instance) => instance.increment,
        r'decrement': (dynamic instance) => instance.decrement,
        r'getCount': (dynamic instance) => instance.getCount,
        r'goTwoLess': (dynamic instance) => instance.goTwoLess
      },
      {},
      null,
      [])
};

final _memberSymbolMap = null;

void initializeReflectable() {
  r.data = _data;
  r.memberSymbolMap = _memberSymbolMap;
}
